model/src/main/java/com/microsoft/java/bs/gradle/model/actions/GetSourceSetsAction.java

```diff
@@ -37,79 +37,144 @@ public class GetSourceSetsAction implements BuildAction<GradleSourceSets> {
-          dependencies.add(new DefaultBuildTargetDependency(otherSourceSet));
-        }
+  private Collection<GradleBuild> fetchIncludedBuilds(BuildController buildController) {
+    Map<String, GradleBuild> builds = new HashMap<>();
+    GradleBuild build = buildController.getBuildModel();
+    String rootProjectName = build.getRootProject().getName();
+    fetchIncludedBuilds(buildController, build, builds, rootProjectName);
+    return builds.values();
+  }
+
+  private void fetchIncludedBuilds(BuildController buildController, GradleBuild build,
+      Map<String, GradleBuild> builds, String rootProjectName) {
+    if (builds.containsKey(rootProjectName)) {
+      return;
+    }
+    builds.put(rootProjectName, build);
+    // Cannot use GradleVersion.current() in BuildAction as that will return the Tooling API version
+    // Cannot use BuildEnvironment to get GradleVersion as that doesn't work pre-3.0 even though
+    // documentation has it added in version 1.
+    // So just handle exceptions
+    Set<? extends GradleBuild> moreBuilds;
+    try {
+      // added in 4.10
+      moreBuilds = build.getEditableBuilds();
+    } catch (Exception e1) {
+      try {
+        // added in 3.3
+        moreBuilds = build.getIncludedBuilds();
+      } catch (Exception e2) {
+        moreBuilds = null;
       }
+    }
+    if (moreBuilds != null) {
+      for (GradleBuild includedBuild : moreBuilds) {
+        String includedBuildName = includedBuild.getRootProject().getName();
+        fetchIncludedBuilds(buildController, includedBuild, builds, includedBuildName);
+      }
+    }
+  }
 
-      DefaultGradleSourceSet sourceSet = new DefaultGradleSourceSet(entry.getKey());
-      sourceSet.setBuildTargetDependencies(dependencies);
-      sourceSets.add(sourceSet);
+  /**
+   * Fetches source sets from the provided Gradle build model.
+   *
+   * @param buildController The Gradle build controller used to interact with the build.
+   * @param builds The Gradle build models representing the build and included builds.
+   */
+  private List<GradleSourceSet> fetchModels(BuildController buildController,
+      Collection<GradleBuild> builds) {
 
+    List<GetSourceSetAction> projectActions = new ArrayList<GetSourceSetAction>();
+    for (GradleBuild build : builds) {
+      for (BasicGradleProject project : build.getProjects()) {
+        projectActions.add(new GetSourceSetAction(project));
+      }
     }
 
-    return new DefaultGradleSourceSets(sourceSets);
+    // since the model returned from Gradle TAPI is a wrapped object, here we re-construct it
+    // via a copy constructorso we can treat as a DefaultGradleSourceSet and
+    // populate source set dependencies.
+    List<GradleSourceSet> sourceSets = buildController.run(projectActions).stream()
+        .flatMap(ss -> ss.getGradleSourceSets().stream())
+        .map(ss -> new DefaultGradleSourceSet(ss))
+        .collect(Collectors.toList());
```

Comment:
Do we still need a copy constructor here @jdneo?

Reply:
Haven't found a time to do the verification. Filed an issue to track: https://github.com/microsoft/build-server-for-gradle/issues/175